compiler/rustc_next_trait_solver/src/solve/assembly/structural_traits.rs

```diff
@@ -304,9 +304,10 @@ pub(in crate::solve) fn extract_tupled_inputs_and_output_from_callable<I: Intern
             let kind_ty = args.kind_ty();
             let sig = args.coroutine_closure_sig().skip_binder();
 
-            let coroutine_ty = if let Some(closure_kind) = kind_ty.to_opt_closure_kind()
-                && !args.tupled_upvars_ty().is_ty_var()
-            {
+            // FIXME: let_chains
+            let kind = kind_ty.to_opt_closure_kind();
+            let coroutine_ty = if kind.is_some() && !args.tupled_upvars_ty().is_ty_var() {
+                let closure_kind = kind.unwrap();
```

Comment:
```suggestion
            let coroutine_ty = if let Some(kind) = kind_ty.to_opt_closure_kind().filter(|_| !args.tupled_upvars_ty().is_ty_var()) {
```

Reply:
I feel like this is so much harder to read